File: #1 Two Sum/two_sum.py
from itertools import combinations
from typing import List

# O(n^2)
def twoSum(nums: List[int], target: int) -> List[int]:
    for j,m in enumerate(nums):
        for i,n in enumerate(nums):
            print(f'{m} {n}')
            if m + n == target and i != j: return [j,i]

# O(n(n-1)/2)
def twoSum(nums: List[int], target: int) -> List[int]:
    for j,m in enumerate(nums):
        for i,n in enumerate(nums[j:]):
            print(f'{m} {n}')
            if m + n == target and j != i+j: return [j,i+j]

# Best solution by blue_sky5
# O(n)
def twoSum(nums: List[int], target: int) -> List[int]:
    seen = {}
    for i,n in enumerate(nums):
        if target - n in seen:
            return [seen[target - n], i]
        seen[n] = i

print(twoSum([-3,4,3,90], 0))

File: #1 Two Sum/two_sum.rs
use std::collections::HashMap;

impl Solution {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
        let mut map: HashMap<i32,usize> = HashMap::new();
        for (i,n) in nums.into_iter().enumerate() {
            match map.get(&(target-n)) {
                None => {map.insert(n, i);},
                Some(&j) => return vec![j as i32, i as i32],
            }
        }
        vec![]
    }
}
